### ISSUE TITLE
[jvm-packages] xgboost4j-spark: bug for persistence of XGBoostEstimator

### ISSUE THREAD

**Issue by YCG09** [dmlc/xgboost#2717]
Operating System: Centos7.0 (with Spark 2.1.1)
Package used: jvm, xgboost4j-spark
`xgboost` version used: 0.7 (May 9th)

I have used xgboost4j-spark for text multi-classification for the past time, it works well. But recently I found a bug when saving a model on hdfs and reloading it afterwards, I just got binary-classification result which is unreasonable. 

When I build the model and use it as below, 

`val paramMap = new mutable.HashMap[String, Any]()`
`paramMap += "nworkers" -> 12`
`paramMap += "use_external_memory" -> false`
`paramMap += "num_round" -> 80`
`paramMap += "objective" -> "multi:softprob"`
`paramMap += "num_class" -> 50`
`val xgboostModel = new XGBoostEstimator(paramMap.toMap).setLabelCol("label").setFeaturesCol("features").setPredictionCol("prediction")`
`val pipeline = new Pipeline().setStages(Array(assembler, xgboostModel))`
`val pipelineModel = pipeline.fit(trainingData)`
`val prediction = pipelineModel.transform(testData)`
`prediction.select("probabilities","prediction").show()`
`prediction.select("probabilities","prediction").take(1)`

I got this, it's correct, there are 50 classes

`+--------------------+----------+`
`|       probabilities|prediction|`
`+--------------------+----------+`
`|[0.01851459778845...|       3.0|`
`|[0.02564118243753...|       6.0|`
`|[0.01733866706490...|       2.0|`
`|[0.02113393507897...|      22.0|`
`|[0.01750198006629...|      21.0|`
`|[0.02301015704870...|      18.0|`
`|[0.02395999617874...|      21.0|`
`|[0.01945679262280...|       1.0|`
`|[0.02109740488231...|      16.0|`
`|[0.02237698808312...|      22.0|`
`|[0.01965968124568...|       1.0|`
`|[0.02031336911022...|       9.0|`
`|[0.01754862070083...|       1.0|`
`|[0.02307951636612...|      18.0|`
`|[0.02055678702890...|       1.0|`
`|[0.02114429697394...|       5.0|`
`|[0.02157610096037...|      15.0|`
`|[0.01917697675526...|      11.0|`
`|[0.02035001106560...|       3.0|`
`|[0.01709185913205...|       2.0|`
`+--------------------+----------+`

`[[0.018514597788453102,0.024826588109135628,0.018507210537791252,0.056314416229724884,0.019766518846154213,0.018494458869099617,0.018510639667510986,0.018490618094801903,0.01955214887857437,0.018467290326952934,0.018479108810424805,0.020859576761722565,0.01847022958099842,0.019066812470555305,0.020105114206671715,0.01852237619459629,0.021153058856725693,0.018469858914613724,0.01853814721107483,0.018480023369193077,0.01942853443324566,0.01851056143641472,0.01848248951137066,0.018485138192772865,0.01848122663795948,0.019552918151021004,0.0184804517775774,0.018460988998413086,0.018459614366292953,0.01960429735481739,0.018465086817741394,0.019087765365839005,0.018509963527321815,0.018484629690647125,0.018471864983439445,0.01846720278263092,0.01...`

but when I save the pipeline model on the hdfs and reuse it like this:

`pipelineModel.write.save("...")`
`val pipelineModelTrained = sparkSession.sparkContext.broadcast(PipelineModel.read.load("..."))`
`val prediction = pipelineModelTrained.value.transform(testData)`
`prediction.select("probabilities","prediction").show()`
`prediction.select("probabilities","prediction").take(1)`

I got the result below:

`+--------------------+----------+`
`|       probabilities|prediction|`
`+--------------------+----------+`
`|[0.98148542642593...|       0.0|`
`|[0.97435879707336...|       0.0|`
`|[0.98266130685806...|       0.0|`
`|[0.97886604070663...|       0.0|`
`|[0.98249804973602...|       0.0|`
`|[0.97698986530303...|       0.0|`
`|[0.97604000568389...|       0.0|`
`|[0.98054319620132...|       0.0|`
`|[0.97890257835388...|       0.0|`
`|[0.97762298583984...|       0.0|`
`|[0.98034030199050...|       0.0|`
`|[0.97968661785125...|       0.0|`
`|[0.98245137929916...|       0.0|`
`|[0.97692048549652...|       0.0|`
`|[0.97944319248199...|       0.0|`
`|[0.97885572910308...|       0.0|`
`|[0.97842389345169...|       0.0|`
`|[0.98082304000854...|       0.0|`
`|[0.97964996099472...|       0.0|`
`|[0.98290812969207...|       0.0|`
`+--------------------+----------+`

`[[0.9814854264259338,0.018514597788453102],0.0]`

it‘s strange, I just get binary-classificatin result, and the prediction are all 0.0. Then, I attempt to save XGBoostEstimator only instead of pipeline and reuse it like this:

`val xgboostStage = pipelineModel.stages(1).asInstanceOf[XGBoostClassificationModel]`
`xgboostStage.write.save("...")`
`val xgboostModel = XGBoostModel.read.load("...")`
`val prediction = xgboostModel.transform(testData)`
`prediction.select("probabilities","prediction").show()`
`prediction.select("probabilities","prediction").take(1)`

But I got the same problem......then I attempt to use `multi:softmax` instead of `multi:softprob`, if model is not saved to hdfs it's got this:

`+-------------+----------+`
`|probabilities|prediction|`
`+-------------+----------+`
`|        [3.0]       |       0.0   |`
`|        [6.0]       |       0.0   |`
`|        [2.0]       |       0.0   |`
`|       [22.0]      |       0.0   |`
`|       [21.0]      |       0.0   |`
`|       [18.0]      |       0.0   |`
`|       [21.0]      |       0.0   |`
`|        [1.0]       |       0.0   |`
`|       [16.0]      |       0.0   |`
`|       [22.0]      |       0.0   |`
`|        [1.0]       |       0.0   |`
`|        [9.0]       |       0.0   |`
`|        [1.0]       |       0.0   |`
`|       [18.0]      |       0.0   |`
`|        [1.0]       |       0.0   |`
`|        [5.0]       |       0.0   |`
`|       [15.0]      |       0.0   |`
`|       [11.0]      |       0.0   |`
`|        [3.0]       |       0.0   |`
`|        [2.0]       |       0.0   |`
`+-------------+----------+`

I don't kown why `setPredictionCol("prediction")` doesn't take effect, but `probabilities` col is the correct result which should be the prediction label. Moreover, when I save the pipeline model to the hdfs and reuse it, I got this:

`+-------------+----------+`
`|probabilities |prediction|`
`+-------------+----------+`
`|   [-2.0,3.0]    |      1.0      |`
`|   [-5.0,6.0]    |      1.0      |`
`|   [-1.0,2.0]    |      1.0      |`
`| [-21.0,22.0]  |      1.0      |`
`| [-20.0,21.0]  |      1.0      |`
`| [-17.0,18.0]  |      1.0      |`
`| [-20.0,21.0]  |      1.0      |`
`|    [0.0,1.0]    |      1.0      |`
`| [-15.0,16.0]  |      1.0      |`
`| [-21.0,22.0]  |       1.0     |`
`|    [0.0,1.0]    |       1.0     |`
`|   [-8.0,9.0]    |       1.0     |`
`|    [0.0,1.0]    |       1.0     |`
`| [-17.0,18.0]  |      1.0      |`
`|    [0.0,1.0]    |       1.0     |`
`|   [-4.0,5.0]    |      1.0      |`
`| [-14.0,15.0]  |      1.0      |`
`| [-10.0,11.0]  |      1.0      |`
`|   [-2.0,3.0]    |      1.0      |`
`|   [-1.0,2.0]    |      1.0      |`
`+-------------+----------+`

It's also strange, and I found that the second column of  `probabilities` is the prediction result, and I got all the same result when I only save XGBoostEstimator model.

I think it's a bug for persistence of XGBoostEstimator, can anyone here help me to solve this problem? Thanks.

**Comment by YCG09** [comment]
@CodingCat Do you have any idea?

**Comment by CodingCat** [comment]
see https://github.com/dmlc/xgboost/pull/2172?

**Comment by YCG09** [comment]
@CodingCat I have read the solution of #2172. It is the method for saving and reloading multi-classification alone, which is only a part of my pipeline model. How can I save and reload pipeline model, and achieve multi-classification meawhile.

**Comment by YCG09** [comment]
@CodingCat @superbobry I'have already check problom #2172,  but I still don't how to settle the problom I proposed, can you explain in more detail?

**Comment by superbobry** [comment]
@YCG09 could you try building #2172 and let us know if it works for you?

**Comment by YCG09** [comment]
@superbobry Ok, I will have a try.

**Comment by YCG09** [comment]
@superbobry Thanks! #2742 can solve my problem, pipeline model also works well.
